content/blog/2022-03-31-training-and-saving-models-on-a-dedicated-runner.md

```diff
@@ -0,0 +1,380 @@
+
+> 💡 The name of the workflow doesn’t matter, as long as it’s a `.yaml` and
+> located in the `.github/workflows` directory. You can have multiple workflows
+> in there as well.
+> ([Documentation](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions))
+
+The workflow we defined first provisions a runner on AWS and then uses that
+runner to train the model. Once we have the outputs saved, we want to export
+them out of the runner. That way we can simply discard the runner once we are
+done training.
+
+In this guide we will explore two options to export the model from the runner:
+
+1. Push the model directly to a git repository;
+2. Push the model to a DVC remote and reference that file in your git
+   repository.
+
+In both cases, the other files (`confusion_matrix.png` etc.) will be pushed
+directly to the git repository. All of this is done in a dedicated `experiment`
+branch and a merge request is automatically created.
+
+## 1. Push the model directly to a git repository
+
+If the model is small we might want to commit it to our repository
+automatically. The following Github workflow deploys a runner on AWS, generates
+some data, trains and saves a model (see `train.py`), pushes the results to a
+new experiment branch, and creates a merge request for those results.
+
+Using `cml pr "."` is the command that takes all of our files, pushes them to a
+new branch, and creates a merge request. Because we saved the model as a binary
+in `model/random_forest.joblib` this file is included in the merge request.
+
+```yaml
+name: CML
+on: # Here we use two triggers; on push and daily at 08:00
+  push:
+  schedule:
+    - cron: '0 8 * * *'
+jobs:
+  deploy-runner:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: iterative/setup-cml@v1
+      - uses: actions/checkout@v2
+      - name: Deploy runner on EC2
+        env:
+          REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+        run: |
+          cml runner \
+              --cloud=aws \
+              --cloud-region=eu-west \
+              --cloud-type=m \
+              --cloud-gpu=nogpu \
+              --single
+  train-model:
+    needs: deploy-runner
+    runs-on: [self-hosted]
+    timeout-minutes: 120 # 2h
+    container:
+      image: docker://iterativeai/cml:0-dvc2-base1
+    steps:
+      - uses: actions/checkout@v2
+      - name: Train model
+        env:
+          REPO_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+        run: |
+          pip install -r requirements.txt
+          python train.py
+
+          # Create CML report
+          cat model/metrics.txt >> model/report.md
+          cml publish model/confusion_matrix.png --md >> model/report.md
+          cml send-comment model/report.md
+
+          # Create pull request
+          cml pr "."
```

Comment:
I'm assuming:

- don't want `report.md` committed
- want PR (rather than commit) to have a report associated
- want report to be updated when subsequent commits are pushed

```suggestion
          cml ci
          pip install -r requirements.txt
          python train.py

          # Create pull request
          cml pr .

          # Create CML report
          cat model/metrics.txt >> model/report.md
          cml publish model/confusion_matrix.png --md >> model/report.md
          cml send-comment --pr --update model/report.md
```

Reply:
Thanks for the feedback, @casperdcl!